### PR TITLE
Re-enable and improve field autocompletes

### PR DESCRIPTION
**Pull request by jmuharsky**
* Adds an optional attribute popupboxDisplayName {string}, which will use the specified property name instead of the selected value, if provided.  This reduces the need for custom templates to handle different data shapes.
* Adds PicklistService, which now houses the picklists and PicklistStates (EMPTY, ERROR, LOADING, LOADED) to track state.
* Adds an optional bound attribute popupboxState {ngExpression.<string>}.  If provided, the resulting PicklistState will be used to show loading, error and empty sections appropriately.

### CODE REVIEW

**Comment by cmccoy** [comment]
Any chance of cleaning up this history?  83c21cc02c012f20f0c7d4a8981dddbec712a5da seems like it should be a separate PR, and maybe overlaps with #125.

**Comment by jmuharsky** [comment]
History cleaned up.  PTAL?

File: client/components/popupbox/popupbox-directive.js [hunk 1]
```diff
@@ -121,19 +127,33 @@ explorer.components.popupbox.PopupboxDirective = function($timeout) {
...
       };
 
       /**
-       * When a value is selected, sets the popupboxModel and closes the popup.
-       * @param {*} value The value that should be selected.
+       * Returns the display text for a row of data.  If popupbox-display-attr
+       * is specified, the matching property will be returned.  Otherwise, the
+       * entire object will be returned.
+       *
+       * @param {*} data The data that should be returned.
```
Comment (cmccoy) [inline client/components/popupbox/popupbox-directive.js]: `@return`?

File: client/components/popupbox/popupbox-directive.js [hunk 2]
```diff
@@ -183,12 +207,19 @@ explorer.components.popupbox.PopupboxDirective = function($timeout) {
...
       };
 
+      /**
+       * Returns true if the displayValue starts with the current text,
+       * otherwise false.
+       */
+      scope.startsWith = function(data) {
+        let displayValue = scope.getDisplayValue(data);
+
+        return (displayValue.indexOf(input[0].value) == 0);
```
Comment (cmccoy) [inline client/components/popupbox/popupbox-directive.js]: Consider [goog.string.startsWith](http://closure-library.googlecode.com/git-history/docs/namespace_goog_string.html) for clarity.

File: client/components/widget/query/picklist/picklist-service.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  let picklist = this.picklists[picklistName];
+
+  if (picklist.state === PicklistStates.LOADING) {
+    return;
+  }
+
+  picklist.state = PicklistStates.LOADING;
+  let promise = this.dataSvc.list(picklistName, queryFilter);
+
+  promise.then(angular.bind(this, function(picklistData) {
```
Comment (cmccoy) [inline client/components/widget/query/picklist/picklist-service.js]: Nit: Recommend against mixing `angular.bind` and arrow functions.

**Comment by cmccoy** [comment]
Couple small things; LGTM.

**Comment by cmccoy** [comment]
(and thanks for cleaning up the commit history!)

File: client/components/popupbox/popupbox-directive.js (hunk 1 above)
Comment (jmuharsky) [inline client/components/popupbox/popupbox-directive.js]: Done.

File: client/components/popupbox/popupbox-directive.js (hunk 2 above)
Comment (jmuharsky) [inline client/components/popupbox/popupbox-directive.js]: Done.

**Pull request merged**